pyccel/parser/semantic.py

```diff
@@ -3970,90 +3973,129 @@ def _visit_For(self, expr):
-                if class_type.rank == 0:
-                    dvar['shape'] = None
-                    dvar['memory_handling'] = 'stack'
-                var  = Variable(class_type, var, **dvar)
+            elif isinstance(a, PythonEnumerate):
+                var_name1 = self.scope.get_expected_name(expr.indices[i][0])
+                var_name2 = self.scope.get_expected_name(expr.indices[i][1])
+                variables.append(handle_int_loop_variable(var_name1, body.scope))
+                variables.append(handle_iterable_variable(var_name2, a.element, body.scope))
                 stop = a.element.shape[0]
 
-            elif isinstance(a, Variable):
-                dvar  = self._infer_type(a)
-                class_type = dvar.pop('class_type')
-                if class_type.rank > 0:
-                    class_type = class_type.switch_rank(class_type.rank-1)
-                    dvar['shape'] = (dvar['shape'])[1:]
-                if class_type.rank == 0:
-                    dvar['shape'] = None
-                    dvar['memory_handling'] = 'stack'
-                var  = Variable(class_type, var, **dvar)
-                stop = a.shape[0]
+            elif isinstance(a, PythonZip):
+                for idx, arg in enumerate(a.args):
+                    var = self.scope.get_expected_name(expr.indices[i][idx])
+                    variables.append(handle_iterable_variable(var, arg, body.scope))
+                stop = a.get_range().stop
+
+            elif isinstance(a, VariableIterator):
+                var = self.scope.get_expected_name(expr.indices[i])
+                variables.append(handle_iterable_variable(var, a.variable, body.scope))
+                stop = a.variable.shape[0]
 
             else:
                 errors.report(PYCCEL_RESTRICTION_TODO,
                               bounding_box=(self.current_ast_node.lineno, self.current_ast_node.col_offset),
                               severity='fatal')
-            existing_var = self.scope.find(var.name, 'variables')
-            if existing_var:
-                if self._infer_type(existing_var) != dvar:
-                    errors.report(f"Variable {var} already exists with different type",
-                            symbol = expr, severity='error')
-            else:
-                self.scope.insert_variable(var)
-            step.invalidate_node()
+
+            for var, _ in variables:
+                existing_var = self.scope.find(var.name, 'variables')
+                if var.name == expr.lhs:
+                    errors.report(f"Variable {var} has the same name as the left hand side",
+                            symbol = expr, severity='fatal')
+                if existing_var or var.name ==  expr.lhs:
+                    if self._infer_type(existing_var)['class_type'] != var.class_type:
+                        return errors.report(f"Variable {var} already exists with different type",
+                                symbol = expr, severity='error')
```

Comment:
Strange to return an error. Should this be fatal or should it be:
```suggestion
                        errors.report(f"Variable {var} already exists with different type",
                                symbol = expr, severity='error')
```
?

Reply:
You are right! it should be fatal! like in this case:
```python
    i = 5
    a = [3.4]
    b = [i-1 for i in a]
```